os/logm/logm_process.c

```diff
@@ -17,39 +17,96 @@
  ****************************************************************************/
 #include <unistd.h>
 #include <stdio.h>
+#include <stdlib.h>
+#include <sys/types.h>
+#include <arch/irq.h>
 #include <tinyara/logm.h>
+#include <tinyara/config.h>
 #include "logm.h"
 
-volatile int g_logm_isready;
-extern char g_logm_rsvbuf[][LOGM_MAX_MSG_LEN];
-extern int g_logm_head;
-extern int g_logm_tail;
-extern int g_logm_count;
+volatile int logm_bufsize = LOGM_BUFFER_SIZE;
+volatile int logm_print_interval = LOGM_PRINT_INTERVAL * 1000;
+char *volatile g_logm_rsvbuf = NULL;
+volatile uint8_t logm_status;
 
-int logm_task(int argc, char *argv[])
+static int logm_change_bufsize(void)
 {
-	(void)argc;
-	(void)argv;
+	int buflen;
+
+	/* Get new values */
+	buflen = new_logm_bufsize;
+
+	/* Keep using old size because of invalid parameter */
+	if (buflen < 0) {
+		LOGM_STATUS_CLEAR(LOGM_BUFFER_RESIZE_REQ);
+		return 0;
+	}
+
+	/* Realloc new buffer with new length */
+	g_logm_rsvbuf = (char *)realloc(g_logm_rsvbuf, buflen);
+	if (!g_logm_rsvbuf) {
+		return -1;
+	}
+	memset(g_logm_rsvbuf, 0, buflen);
+	lmdbg("logm new malloced address is 0x%x\n", g_logm_rsvbuf);
+
+	/* Reinitialize all  */
 	g_logm_head = 0;
 	g_logm_tail = 0;
-	g_logm_count = 0;
+	logm_bufsize = buflen;
+	g_logm_available = buflen;
+	g_logm_enqueued_count = 0;
+	g_logm_dropmsg_count = 0;
+	g_logm_overflow_offset = -1;
+
+	LOGM_STATUS_CLEAR(LOGM_BUFFER_RESIZE_REQ);	// finish
+
+	return 0;
+}
+
+int logm_task(int argc, char *argv[])
+{
+	int ret = 0;
+	irqstate_t flags;
+
+	g_logm_rsvbuf = (char *)malloc(logm_bufsize);
+	memset(g_logm_rsvbuf, 0, logm_bufsize);
+
 	/* Now logm is ready */
-	g_logm_isready = 1;
+	LOGM_STATUS_SET(LOGM_READY);
+	g_logm_available = logm_bufsize;
 
 #ifdef CONFIG_LOGM_TEST
 	logmtest_init();
```

Comment:
Some of variables are initialized with definition and others are set as 0 because of global.
currently logm init has logm_start function to run logm task only.
I thought that it is not good to insert this to that.